pkg/synchromanager/clustersynchro/discovery/discovery_cache.go

```diff
@@ -0,0 +1,443 @@
+	// k8s.io/apimachinery/pkg/api/meta/restmapper.go#coerceResourceForMatching
+	gr.Resource = strings.ToLower(gr.Resource)
+	if plural := c.singularToPlural[gr]; !plural.Empty() {
+		gr = plural
+	}
+
+	return c.apiResources[gr], c.resourceVersions[gr]
+}
+
+func (c *DynamicDiscoveryManager) GetAPIResourceAndVersions(resource schema.GroupResource) (*metav1.APIResource, []string) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	var refetchFunc func() (updated bool)
+	if c.customResourceGroups.Has(resource.Group) {
+		// not refresh custom resource
+	} else if c.aggregatorGroups.Has(resource.Group) {
+		refetchFunc = func() bool {
+			// TODO(iceber): Add limit on fetching aggreegator groups
+			// For example, the same group can only be fetched once in 3 minutes and will be fetched again after 3 minutes
+			updated, _ := c.refetchGroupsLocked(sets.NewString(resource.Group))
+			return updated
+		}
+	} else if _, ok := c.groupVersions[resource.Group]; ok {
+		refetchFunc = func() bool {
+			// TODO(iceber): Add limit on fetching server version
+			updated, err := c.fetchServerVersion()
+			if err != nil {
+				utilruntime.HandleError(err)
+			}
+			if !updated {
+				return false
+			}
+
+			updated, _ = c.refetchAllGroupsLocked()
+			return updated
+		}
+	} else {
+		return nil, nil
+	}
+
+	if apiResource, versions := c.getAPIResourceAndVersionsLocked(resource); len(versions) != 0 {
+		return &apiResource, versions
+	}
+
+	if refetchFunc != nil && !refetchFunc() {
+		return nil, nil
+	}
+
+	if apiResource, versions := c.getAPIResourceAndVersionsLocked(resource); len(versions) != 0 {
+		return &apiResource, versions
+	}
+	return nil, nil
+}
+
+func (c *DynamicDiscoveryManager) GetAllResourcesAsSyncResources() []clusterv1alpha2.ClusterGroupResources {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	sortedResources := make([]schema.GroupResource, 0, len(c.resourceVersions))
+	for gr := range c.resourceVersions {
+		sortedResources = append(sortedResources, gr)
+	}
+	sortGroupResources(sortedResources)
+
+	syncResources := make([]clusterv1alpha2.ClusterGroupResources, 0, len(sortedResources))
+	for _, gr := range sortedResources {
+		resource := clusterv1alpha2.ClusterGroupResources{
+			Group:     gr.Group,
+			Resources: []string{gr.Resource},
+			Versions:  c.resourceVersions[gr],
+		}
+		syncResources = append(syncResources, resource)
+	}
+	return syncResources
+}
+
+func (c *DynamicDiscoveryManager) AttachAllCustomResourcesToSyncResources(resources []clusterv1alpha2.ClusterGroupResources) []clusterv1alpha2.ClusterGroupResources {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	sortedCustomResources := make([]schema.GroupResource, 0, len(c.customResourceGroups))
+	for gr := range c.resourceVersions {
+		if !c.customResourceGroups.Has(gr.Group) {
+			sortedCustomResources = append(sortedCustomResources, gr)
+		}
+	}
+	sortGroupResources(sortedCustomResources)
```

Comment:
`!c.customResourceGroups.Has`  If it is included, it is a custom resource

Reply:
you are right